### PR TITLE
store/tikv: update copRanges.slice().

### PR DESCRIPTION
**Pull request by disksing**
I have met panic during test once recently, but not able to reproduce it. The stack shows the problem should be some place inside `buildCopTasks`.
This PR aims to make `copRanges` more robust.
@shenli @zimulala @tiancaiamao

### CODE REVIEW

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -173,7 +173,12 @@ func (r *copRanges) slice(from, to int) *copRanges {
 	if to <= len(r.mid) {
 		ran.mid = r.mid[from:to]
 	} else {
-		ran.mid, ran.last = r.mid[from:], r.last
+		if from <= len(r.mid) {
```
Comment (shenli) [inline store/tikv/coprocessor.go]: Any test case cover this from > len(r.mid)？

Comment (disksing) [inline store/tikv/coprocessor.go]: Yes, the new test iterates all possible combinations of from and to and make sure its behaviour is same as raw slice.

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -173,7 +173,12 @@ func (r *copRanges) slice(from, to int) *copRanges {
...
+			ran.mid = r.mid[from:]
+		}
+		if from < to {
```
Comment (shenli) [inline store/tikv/coprocessor.go]: How about from > len(r.mid) + 1?

Comment (disksing) [inline store/tikv/coprocessor.go]: Then it will be incorrect. `from < to` is specially for the case `from == to`, when the result should be empty slice.

**Comment by shenli** [comment]
LGTM
@coocood @tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM

btw, when slice out of range,  I think panic or return error maybe better than adjust the range

**Comment by disksing** [comment]
@tiancaiamao In golang, it's ok to use len(slice) to do slice.
```
a := []int{1, 2, 3}
b := a[3:3] // Not panic, but got an empty slice.
```

**Pull request merged**